plugins/woocommerce/src/Internal/Admin/ProductDownloadsPreview.php

```diff
@@ -0,0 +1,182 @@
+			$namespace,
+			$route_base . $route_pattern,
+			array(
+				'methods'             => WP_REST_Server::READABLE,
+				'callback'            => array( $this, 'get_preview' ),
+				'permission_callback' => array( $this, 'get_preview_permissions_check' ),
+				'args'                => $args,
+			)
+		);
+	}
+
+	/**
+	 * Permission check for REST API endpoint.
+	 *
+	 * @since 9.9.0
+	 * @param \WP_REST_Request $request Request details.
+	 * @return bool|\WP_Error
+	 */
+	public function get_preview_permissions_check( $request ) {
+		if ( ! current_user_can( 'manage_woocommerce' ) ) {
+			return new WP_Error(
+				'woocommerce_rest_unauthorized',
+				__( 'Unauthorized access.', 'woocommerce' ),
+				array( 'status' => 403 )
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Serve the preview image
+	 *
+	 * @since 9.9.0
+	 * @param \WP_REST_Request $request Request details.
+	 * @return \WP_REST_Response|\WP_Error
+	 */
+	public function get_preview( $request ) {
+		$attachment_id  = $request['attachment_id'];
+		$product_id     = $request['product_id'];
+		$requested_size = $request['size'] ?? 'large';
+
+		$file_path = get_attached_file( $attachment_id );
+		if ( ! $file_path || ! is_readable( $file_path ) ) {
+			return new WP_Error(
+				'woocommerce_rest_file_not_found',
+				__( 'File not found', 'woocommerce' ),
+				array( 'status' => 404 )
+			);
+		}
+
+		$mime_type          = get_post_mime_type( $attachment_id );
+		$allowed_mime_types = array(
+			'image/jpeg',
+			'image/jpg',
+			'image/png',
+			'image/gif',
+			'image/webp',
+		);
+
+		if ( ! in_array( $mime_type, $allowed_mime_types, true ) ) {
+			return new WP_Error(
+				'woocommerce_rest_invalid_file_type',
+				__( 'Invalid file type', 'woocommerce' ),
+				array( 'status' => 403 )
+			);
+		}
+
+		$size = 'full' === $requested_size ? 'large' : $requested_size;
+
+		$resized = image_get_intermediate_size( $attachment_id, $size );
+		if ( $resized && isset( $resized['path'] ) ) {
+			$uploads_dir       = wp_upload_dir();
+			$resized_file_path = $uploads_dir['basedir'] . '/' . $resized['path'];
+			if ( is_readable( $resized_file_path ) ) {
+				$file_path = $resized_file_path;
+			}
+		}
+
+		// Clean any existing output.
+		ob_clean();
+
+		nocache_headers();
+		header( 'Content-Type: ' . $mime_type );
+		header( 'Content-Length: ' . filesize( $file_path ) );
+		header( 'Content-Disposition: inline; filename="' . basename( $file_path ) . '"' );
+
+		// We need to use readfile here for binary file output.
+		// phpcs:ignore WordPress.WP.AlternativeFunctions.file_system_operations_readfile
+		readfile( $file_path );
+		exit;
+	}
+
+	/**
+	 * Get secure URL for admin image
+	 *
+	 * @since 9.9.0
+	 * @param int    $product_id    Product ID.
+	 * @param int    $attachment_id Attachment ID.
+	 * @param string $size          Image size.
+	 * @return string Secure admin image URL.
+	 */
+	public function get_admin_image_src_url( $product_id, $attachment_id, $size ) {
```

Comment:
There are lots of spots like this where type-hints could be handy. Or else, we could explicitly cast to `int` (or otherwise sanitize) before placing the values in the `$url` string (just as a security-in-depth measure).

Reply:
I've updatet he code to add type hints and a few explicit casts here: c424ff1170c260bed911d5f49572028b21536ded